web/app/Helpers/PageBuilder.php

```diff
@@ -0,0 +1,344 @@
+                $this->parameterArray[$key] = urldecode($value);
+                $this->parameterTypes[$key] = 'GET';
+                $this->allParameters['GET'][$key] = urldecode($value);
+            }
+
+            /** Lastly, POST values don't need as much manipulation. */
+            foreach ($_POST as $key => $value) {
+                /** Convert any CRLF values to LF */
+                $value = str_replace(["\r\n", "\r", "\n"], "\n", $value);
+
+                /** And add them to the bucket. */
+                $this->parameterArray[$key] = $value;
+                $this->parameterTypes[$key] = 'POST';
+                $this->allParameters['POST'][$key] = urldecode($value);
+            }
+        }
+    }
+
+    /**
+     *  Validate the requested template or screen and retrieve the location of it.
+     */
+    public function setTemplate()
+    {
+        /** The ajaxMode property can be set true by the AjaxModuleWebFlowController */
+        if ($this->ajaxMode === true) {
+            /** We can reliably assume that an AJAX module provided a key called moduleName */
+            $template = $this->parameterArray['moduleName'];
+
+            /**
+             * If we didn't get a valid value back (not empty, contains only
+             *  letters, numbers, underscores, and slashes), use the Empty template. (nothing)
+             */
+            if ($template == null || preg_match('/^[a-z0-9_\/]+$/i', $template) != 1) {
+                $template = 'Empty';
+            }
+
+            /**
+             * Otherwise we're going to assume there's a matching template in
+             *  templates/modules to go with the PHP module in php/modules.
+             *
+             * These AJAX-centric modules generate the HTML to be injected back
+             *  into the DOM wholesale.
+             */
+            $this->moduleTemplate = $template;
+
+            $this->findModuleClass($this->moduleTemplate);
+        } else {
+            /**
+             * If we're not making an AJAX call, we're looking for a Screen instead of a template.
+             * Screens are in php/Screens and are the classes that use the
+             * templates in templates/screens.
+             */
+            $template = $this->parameterArray['template'];
+
+            /**
+             * If we didn't get a valid value back (not empty, contains only
+             *  letters, numbers, underscores, and slashes), use the Index screen. (nothing)
+             */
+            if ($template == null || preg_match('/^[a-z0-9_\/]+$/i', $template) != 1) {
+                $template = 'Index';
+            }
+
+            /**
+             * We receive the call on the php file as a path using two
+             * underscores instead of a slash.  We need to pass the actual path to
+             * our next set of tasks.
+             */
+            $template = str_replace('__', '/', $template);
```

Comment:
This is also done for actions. Does it happen anywhere else? (If so, 3+ usages suggests it would be better as a helper function).